semexprs: refactor 'var' parameter argument fixup

zerbina opened the pull request:
## Summary
This is mostly a refactoring and clean up, but it also adjusts the "is
address taken" analysis to no longer ignore l-value conversions and
statement list expressions. The operand to an `addr` operation now also
uses the same (more correct) analysis as the arguments to `var`
parameters.

`jsgen` depends on the `sfAddrTaken` flag in order to know whether a
location needs to support being modified via a "base + index" (the JS
back-end's way of implementing pointers), so this change fixes a
JS code-gen issue. Only in theory however, as the JS code-generator
implements the relevant bits incorrectly -- although in slightly
different way, wrong code is still generated.

## Details
- rename `analyseIfAddressTakenInCall` to `fixVarArgumentsAndAnalyse`,
  refactor and document it, and make it `nkError` aware
- remove the mutability check from `newHiddenAddrTaken` -- this is now
  the responsibility of the callsite
- move the "is address taken" analysis to a separate procedure
- move the `FakeVarParams` constant to the `ast_query` module
- make `semAddrArg` `nkError` aware, fixing `addr x` reporting an
  `expression has no address` error if expression `x` had an error
- fix the "has no address" error reported for  `addr` always suggesting
  the usage of `unsafeAddr`, even if it was already used
- remove leftover and unused `unsafeAddr`-related logic

---
<!-- Note: section break (`---`) onwards is not in CI merge commit -->

As mentioned in the code comments, I don't think `jsgen` depending on `sfAddrTaken` is a good idea. The analysis itself should also not happen as part of the after-call actions, but rather during `sempass2`. Both are things that are better implemented in separate PRs, however.

<!--
Pull Request(PR) Help

Before Merge Ensure:
* title reads like a short changelog line entry
* code includes tests and is documented
* leave the source better than before, but split out big reformats

See contributor (guide)[https://nim-works.github.io/nimskull/contributing.html]
for details, especially if you're new to this project.

Tips that make PRs easier:
* for big/impactful changes, start with chat/discussions to refine ideas
* refine the pull request message over time; don't have to nail it in one go
* handle the single commit message requirement at the end of review

saem commented:
Did a quick skim, I like it. The js gen dependency on `sfAddressTaken` is definitely very suss.

zerbina committed:
semexprs: refactor 'var' parameter argument fixup

## Summary
This is mostly a refactoring and clean up, but it also adjusts the "is
address taken" analysis to no longer ignore l-value conversions and
statement list expressions. The operand to an `addr` operation now also
uses the same (more correct) analysis as the arguments to `var`
parameters.

`jsgen` depends on the `sfAddrTaken` flag in order to know whether a
location needs to support being modified via a "base + index" (the JS
back-end's way of implementing pointers), so this change fixes a
JS code-gen issue. Only in theory however, as the JS code-generator
implements the relevant bits incorrectly -- although in slightly
different way, wrong code is still generated.

## Details
- rename `analyseIfAddressTakenInCall` to `fixVarArgumentsAndAnalyse`,
  refactor and document it, and make it `nkError` aware
- remove the mutability check from `newHiddenAddrTaken` -- this is now
  the responsibility of the callsite
- move the "is address taken" analysis to a separate procedure
- move the `FakeVarParams` constant to the `ast_query` module
- make `semAddrArg` `nkError` aware, fixing `addr x` reporting an
  `expression has no address` error if expression `x` had an error
- fix the "has no address" error reported for  `addr` always suggesting
  the usage of `unsafeAddr`, even if it was already used
- remove leftover and unused `unsafeAddr`-related logic

saem commented:
Bors r+

Thank you!

Inspired by this PR, we should discuss what the semantic meaning of various types, handles, locations all should mean. Maybe a good way to do that would be to attempt to make a table to describe it. Something like:

|scope|modifier|type|meaning|
|-------|------|-------|----|
|parameter|var|*|var applies to the location, arg must be a handle to a variable location|

The dimensions of the table and constiuents would be a good way to get things nailed down, surface unnamed or specified concepts, etc.

saem commented:
bors r+

saem commented:
Bors r+
